pkg/dynamicinstrumentation/codegen/codegen.go

```diff
@@ -128,13 +128,38 @@ func collectLocationExpressions(param *ditypes.Parameter) []ditypes.LocationExpr
 		}
 		queue = append(queue, top.ParameterPieces...)
 		if len(top.LocationExpressions) > 0 {
-			collectedExpressions = append(top.LocationExpressions, collectedExpressions...)
+			expressions := []ditypes.LocationExpression{}
+			for i := range top.LocationExpressions {
+				expressions = append(expressions, collectSubLocationExpressions(top.LocationExpressions[i])...)
+			}
+			collectedExpressions = append(expressions, collectedExpressions...)
 			top.LocationExpressions = []ditypes.LocationExpression{}
 		}
 	}
 	return collectedExpressions
 }
 
+func collectSubLocationExpressions(location ditypes.LocationExpression) []ditypes.LocationExpression {
+	collectedExpressions := []ditypes.LocationExpression{}
+	queue := []ditypes.LocationExpression{location}
+	var top ditypes.LocationExpression
+
+	for {
+		if len(queue) == 0 {
+			break
+		}
+		top = queue[0]
+		queue = queue[1:]
+		queue = append(queue, top.IncludedExpressions...)
+		if top.Opcode != ditypes.OpPopPointerAddress {
+			collectedExpressions = append(collectedExpressions, top)
+		}
+
+		top.IncludedExpressions = []ditypes.LocationExpression{}
```

Comment:
This is a really interesting way to do a fold, I had to stare at it for a second to ensure it terminates so it might be worth a comment. I think `nil` is also valid with the ellipsis notation which might be even more clear.

```suggestion
		// Clear IncludedExpressions to prevent reprocessing and ensure the loop terminates
		top.IncludedExpressions = nil
```

Reply:
The intention here is not to ensure loop termination, it's to prevent the instruction from being use again when generating BPF programs.

Reply:
Oh okay, and we know that eventually `top.IncludedExpressions` will be empty as we loop through the locations 👍